packages/eslint-plugin/src/rules/prefer-optional-chain.ts

```diff
@@ -48,6 +48,59 @@ export default util.createRule({
   create(context) {
     const sourceCode = context.getSourceCode();
     return {
+      'LogicalExpression[operator="||"], LogicalExpression[operator="??"]'(
+        node: TSESTree.LogicalExpression,
+      ): void {
+        const leftNode = node.left;
+        const rightNode = node.right;
+        const parentNode = node.parent;
+        const isRightNodeAnEmptyObjectLiteral =
+          rightNode.type === AST_NODE_TYPES.ObjectExpression &&
+          rightNode.properties.length === 0;
+        // Ignore nodes that evaluate to numbers or booleans
+        const canLeftNodeBeANullishObject =
+          leftNode.type !== AST_NODE_TYPES.BinaryExpression &&
+          leftNode.type !== AST_NODE_TYPES.UnaryExpression;
```

Comment:
I don't think it's worth having this check.
This is a deep rabbit-hole to go down to statically analyse all of the operators to determine if they could possibly return a nullish value.
Only handling a subset of them feels wrong, because it means the rule won't check some cases and will check others - to a user it just feels janky to be inconsistent.

There's also things like `(typeof x as string) || {}` which would be ignored because your code doesn't "look past" the assertion.
Another example of the rabbit-hole - your check will false-negative on `void x` - which is a `UnaryExpression` that always returns `undefined`.

I would just go with the same logic of the rest of the rule - which is to always assume the left side could be nullable.

---

If, however, you wanted to implement something more thorough - there is an example code like this which could be reworked and built on top of
https://github.com/eslint/eslint/blob/97a4f6385cbcd8ea251de17dc2f1683132c2d0d9/lib/rules/no-constant-binary-expression.js#L23-L84

The basic idea would be to do this:
1) if the thing is an `Identifier`, use scope analysis to find the definition
   a) if the definition is variable with more than one assignment, assume it could be nullable.
   b) else analyse the variable's singular written value
2) if the value is statically non-null, then skip it
3) if the value is potentially nullish, continue evaluating.

Reply:
I had good intentions to not show the lint error on things that are clearly messed up, but the rule cannot fix or warn correctly, like values that cannot be objects (accessing properties on a boolean or number just goes to the prototype, and most of them are methods that could better be accessed directly)

I don't think I'm familiar enough with the API of how to perform scope analysis and I assume it's not a trivial change.
If you think it's enough to drop the check I'll go with that approach.

Reply:
Definitely agree with the good intentions!
I think for now we can drop the check for this first commit, and then we can come back and assess to see if we can holistically improve this rule with as many cases as we can.

Reply:
Hi, @bradzacher & @JoshuaKGoldberg Is there anything else I need to change?
I think we can resolve the requested changes.